libraries/chain/controller.cpp

```diff
@@ -1251,19 +1256,38 @@ struct controller_impl {
       }
    } FC_CAPTURE_AND_RETHROW() } /// apply_block
 
+   std::future<block_state_ptr> create_block_state_future( const signed_block_ptr& b ) {
+      EOS_ASSERT( b, block_validate_exception, "null block" );
 
-   void push_block( const signed_block_ptr& b, controller::block_status s ) {
+      auto id = b->id();
+
+      // no reason for a block_state if fork_db already knows about block
+      auto existing = fork_db.get_block( id );
+      EOS_ASSERT( !existing, fork_database_exception, "we already know about this block: ${id}", ("id", id) );
+
+      auto prev = fork_db.get_block( b->previous );
+      EOS_ASSERT( prev, unlinkable_block_exception, "unlinkable block ${id}", ("id", id)("previous", b->previous) );
+
+      return async_thread_pool( [b, prev]() {
+         const bool skip_validate_signee = false;
+         return std::make_shared<block_state>( *prev, move( b ), skip_validate_signee );
+      } );
+   }
+
+   void push_block( std::future<block_state_ptr>& block_state_future ) {
+      controller::block_status s = controller::block_status::complete;
       EOS_ASSERT(!pending, block_validate_exception, "it is not valid to push a block when there is a pending block");
 
       auto reset_prod_light_validation = fc::make_scoped_exit([old_value=trusted_producer_light_validation, this]() {
          trusted_producer_light_validation = old_value;
       });
       try {
-         EOS_ASSERT( b, block_validate_exception, "trying to push empty block" );
-         EOS_ASSERT( s != controller::block_status::incomplete, block_validate_exception, "invalid block status for a completed block" );
+         block_state_ptr new_header_state = block_state_future.get();
+         auto& b = new_header_state->block;
          emit( self.pre_accepted_block, b );
-         bool trust = !conf.force_all_checks && (s == controller::block_status::irreversible || s == controller::block_status::validated);
-         auto new_header_state = fork_db.add( b, trust );
+
+         fork_db.add( new_header_state );
```

Comment:
Calling `fork_database::add( block_state_ptr )` here instead of the original `fork_database::add( signed_block_ptr bool )` skips some checks that were done in the original version.

`create_block_state_future` replicates those `fork_database::add( signed_block_ptr, bool)` checks, but it does so at a different time (and therefore potentially with different state of the fork database) than the call to `fork_db.add` here.

One of the two necessary checks, checking for duplicate blocks, is already handled in `fork_database::add( block_state_ptr )` with the assertion that `inserted.second` is true.

The other check that is necessary to again here (so it could be done in `fork_database::add( block_state_ptr )`) is to ensure that the block still links to an existing block in the fork database. Without that check, it is possible to add a block into the fork database that does not link. For example, a user of `controller` could in theory: 
1. Call `create_block_state_future` on two different blocks A and B where both currently link to existing blocks in the fork database and where B builds off of the current head.
2. Then calls `push_block` with the future of block A which causes: a fork switch, LIB advancement, and purging of orphaned branches which may include the block that the block B builds off of.
3. Then calls `push_block` with the future of block B.

Reply:
The current code does not have more than one `create_block_state_future` in play at a time. It is single threaded from producer plugin `on_incoming_block` to `push_block`. The `create_block_state_future` creation of block_state which includes digest, merkle, and signee verification can run at the same time as `abort_block` but the rest of the order is all single threaded.

Now if we updated the code so that more than one `create_block_state_future` could be in flight at a time then that could certainly be a problem. That would take some work as it does verify previous is in the fork_db. Could create a version of `create_block_state_future` that takes the previous block so that a whole chain of blocks could be processed at once and then verify in fork_db add that the previous is there. However, that would require changing net_plugin to receive more than one block at a time or queue up blocks which would increase latency which is what this is attempting to reduce.

That said the extra check doesn't hurt and should be quick, so I'll go ahead and add it.